### PR TITLE
✨ Added a library for data types of units used within the physical simulation algorithms

### PR DESCRIPTION
**Pull request by Drewniok**
## Description

This PR introduces the unit library ``units`` (https://github.com/nholthaus/units) for all physical simulation algorithms. It assists in writing safer code since it gets harder to accidentally assume the wrong unit of a `double` or `int` for example. 

## Checklist:

- [x] The pull request only contains commits that are related to it.
- [x] I have added appropriate tests and documentation.
- [ ] I have made sure that all CI jobs on GitHub pass.
- [ ] The pull request introduces no new warnings and follows the project's style guidelines.

### CODE REVIEW

File: include/fiction/algorithms/simulation/sidb/critical_temperature.hpp [hunk 1]
```diff
@@ -346,10 +350,10 @@ class critical_temperature_impl
                 break;
             }
 
-            if (std::abs(temp - static_cast<double>(parameter.max_temperature)) < 0.001)
+            if (std::abs((temp - parameter.max_temperature).value()) < 0.001)
```
Comment (marcelwa) [inline include/fiction/algorithms/simulation/sidb/critical_temperature.hpp:353]: Could you use `units::math::abs` here? Would alleviate the need for `.value()`.

File: include/fiction/algorithms/simulation/sidb/critical_temperature.hpp [hunk 2]
```diff
@@ -417,10 +422,10 @@ class critical_temperature_impl
...
```
Comment (marcelwa) [inline include/fiction/algorithms/simulation/sidb/critical_temperature.hpp:425]: See above.

File: include/fiction/algorithms/simulation/sidb/is_ground_state.hpp [hunk 3]
```diff
@@ -41,7 +42,8 @@ template <typename Lyt>
     const auto min_energy_exact  = minimum_energy(exhaustive_results.charge_distributions);
     const auto min_energy_new_ap = minimum_energy(quicksim_results.charge_distributions);
 
-    return std::abs(min_energy_exact - min_energy_new_ap) / min_energy_exact < physical_constants::POP_STABILITY_ERR;
+    return round_to_n_decimal_places((min_energy_exact - min_energy_new_ap).value(), 6) <
```
Comment (marcelwa) [inline include/fiction/algorithms/simulation/sidb/is_ground_state.hpp:45]: Is `std::abs`/`units::math::abs` no longer needed?

File: include/fiction/algorithms/simulation/sidb/occupation_probability_of_excited_states.hpp [hunk 4]
```diff
@@ -20,6 +21,8 @@
 namespace fiction
 {
 
+using units::literals::operator""_K;
```
Comment (marcelwa) [inline include/fiction/algorithms/simulation/sidb/occupation_probability_of_excited_states.hpp:24]: Please add a docstring and RTD documentation.

File: include/fiction/algorithms/simulation/sidb/occupation_probability_of_excited_states.hpp [hunk 5]
```diff
@@ -50,7 +53,7 @@ namespace fiction
     const double partition_function =
         std::accumulate(energy_and_state_type.cbegin(), energy_and_state_type.cend(), 0.0,
                         [&](const double sum, const auto& it)
-                        { return sum + std::exp(-(it.first - min_energy) * 12'000 / temperature); });
+                        { return sum + std::exp(-((it.first - min_energy) * 12'000 / temperature).value()); });
```
Comment (marcelwa) [inline include/fiction/algorithms/simulation/sidb/occupation_probability_of_excited_states.hpp:56]: `units::math::exp` exists

File: include/fiction/algorithms/simulation/sidb/occupation_probability_of_excited_states.hpp [hunk 6]
```diff
@@ -60,7 +63,7 @@ namespace fiction
     {
         if (!state_transparent_erroneous)
         {
-            p += std::exp(-(energies - min_energy) * 12'000 / temperature);
+            p += std::exp(-((energies - min_energy) * 12'000 / temperature).value());
```
Comment (marcelwa) [inline include/fiction/algorithms/simulation/sidb/occupation_probability_of_excited_states.hpp:66]: See above

File: include/fiction/algorithms/simulation/sidb/occupation_probability_of_excited_states.hpp [hunk 7]
```diff
@@ -94,21 +98,21 @@ namespace fiction
     const double partition_function =
         std::accumulate(energy_distribution.cbegin(), energy_distribution.cend(), 0.0,
                         [&](const double sum, const auto& it)
-                        { return sum + std::exp(-(it.first - min_energy) * 12'000 / temperature); });
+                        { return sum + std::exp(-((it.first - min_energy) * 12'000 / temperature).value()); });
```
Comment (marcelwa) [inline include/fiction/algorithms/simulation/sidb/occupation_probability_of_excited_states.hpp:101]: See above

File: include/fiction/algorithms/simulation/sidb/occupation_probability_of_excited_states.hpp [hunk 8]
```diff
@@ -94,21 +98,21 @@ namespace fiction
...
-                        });
+    const double p = std::accumulate(
+        energy_distribution.cbegin(), energy_distribution.cend(), 0.0,
+        [&](const double sum, const auto& it)
+        {
+            // round the energy value of the given valid_layout to six decimal places to overcome
+            // possible rounding errors and for comparability with the min_energy.
+            if (round_to_n_decimal_places(it.first.value(), 6) != round_to_n_decimal_places(min_energy.value(), 6))
+            {
+                return sum + std::exp(-((it.first - min_energy) * 12'000 / temperature).value());
```
Comment (marcelwa) [inline include/fiction/algorithms/simulation/sidb/occupation_probability_of_excited_states.hpp:112]: See above

File: include/fiction/algorithms/simulation/sidb/sidb_simulation_parameters.hpp [hunk 9]
```diff
@@ -5,14 +5,20 @@
...
 #include <cstdint>
 
+#include <units.h>
+
 namespace fiction
 {
 
+using units::literals::operator""_eV;
+using units::literals::operator""_nm;
+using units::literals::operator""_angstrom;
```
Comment (marcelwa) [inline include/fiction/algorithms/simulation/sidb/sidb_simulation_parameters.hpp:20]: Please add docstrings and RTD documentation.

Also, why are the unit literals defined in different places?

File: include/fiction/algorithms/simulation/sidb/sidb_simulation_parameters.hpp [hunk 10]
```diff
@@ -53,35 +62,37 @@ struct sidb_simulation_parameters
...
+    units::length::angstrom_t lat_c;
     /**
      * epsilon_r is the electric permittivity. It is a material specific number (unit-less).
      */
     double epsilon_r;
     /**
      * k is the Coulomb constant and is inversely proportional to the electric permittivity (unit: SI).
      */
-    double k;
+    units::unit_t<units::compound_unit<units::force::newtons, units::area::square_meter,
```
Comment (marcelwa) [inline include/fiction/algorithms/simulation/sidb/sidb_simulation_parameters.hpp:81]: Maybe create an alias for this unit type?

File: include/fiction/technology/charge_distribution_surface.hpp [hunk 11]
```diff
@@ -340,7 +343,11 @@ class charge_distribution_surface<Lyt, false> : public Lyt
             {
                 if (const auto local_pot = this->get_local_potential(c); local_pot.has_value())
                 {
-                    if (-*local_pot + strg->phys_params.mu < -physical_constants::POP_STABILITY_ERR)
+                    if ((units::energy::electron_volt_t(
+                             units::convert<units::energy::joule, units::energy::electron_volt>(
+                                 (-*local_pot * units::constants::e).value())) +
+                         strg->phys_params.mu)
+                            .value() < -physical_constants::POP_STABILITY_ERR)
```
Comment (marcelwa) [inline include/fiction/technology/charge_distribution_surface.hpp:350]: Maybe add a comment on what's happening here?

File: include/fiction/technology/charge_distribution_surface.hpp [hunk 12]
```diff
@@ -549,42 +558,49 @@ class charge_distribution_surface<Lyt, false> : public Lyt
...
+        // convert<square_root<square_kilometer>, meter>(1.0);
+        strg->system_energy =
+            units::energy::electron_volt_t(units::convert<units::energy::joule, units::energy::electron_volt>(
+                (total_potential * units::constants::e).value()));
 
-        strg->system_energy = total_energy;
+        // strg->system_energy =(units::convert<units::energy::joule, units::energy::electron_volt>(total_potential *
+        // units::constants::e));
+
+        //     strg->system_energy = units::energy::electron_volt_t((total_potential).value());
```
Comment (marcelwa) [inline include/fiction/technology/charge_distribution_surface.hpp:596]: Please remove the commented-out code.

File: include/fiction/technology/charge_distribution_surface.hpp [hunk 13]
```diff
@@ -549,42 +558,49 @@ class charge_distribution_surface<Lyt, false> : public Lyt
...
     {
-        double total_energy = 0;
+        units::voltage::volt_t total_potential(0.0);
 
         for (uint64_t i = 0; i < strg->loc_pot.size(); ++i)
         {
-            total_energy += 0.5 * strg->loc_pot[i] * charge_state_to_sign(strg->cell_charge[i]);
+            total_potential += 0.5 * strg->loc_pot[i] * charge_state_to_sign(strg->cell_charge[i]);
         }
+        // convert<square_root<square_kilometer>, meter>(1.0);
```
Comment (marcelwa) [inline include/fiction/technology/charge_distribution_surface.hpp:588]: Please remove the commented-out code.

File: include/fiction/technology/charge_distribution_surface.hpp [hunk 14]
```diff
@@ -609,12 +625,16 @@ class charge_distribution_surface<Lyt, false> : public Lyt
...
-                           ((-it + strg->phys_params.mu_p) > -physical_constants::POP_STABILITY_ERR)) ||
+                           ((units::energy::electron_volt_t(-it.value()) + strg->phys_params.mu_p).value() >
+                            -physical_constants::POP_STABILITY_ERR)) ||
                           ((strg->cell_charge[for_loop_counter] == sidb_charge_state::NEUTRAL) &&
-                           ((-it + strg->phys_params.mu) > -physical_constants::POP_STABILITY_ERR) &&
-                           (-it + strg->phys_params.mu_p) < physical_constants::POP_STABILITY_ERR));
+                           ((units::energy::electron_volt_t(-it.value()) + strg->phys_params.mu).value() >
+                            -physical_constants::POP_STABILITY_ERR) &&
+                           (units::energy::electron_volt_t(-it.value()) + strg->phys_params.mu_p).value() <
+                               physical_constants::POP_STABILITY_ERR));
```
Comment (marcelwa) [inline include/fiction/technology/charge_distribution_surface.hpp:637]: Are the `.value()` calls needed here? Again, I'm not yet familiar with the library and this is an honest question.

File: include/fiction/technology/charge_distribution_surface.hpp [hunk 15]
```diff
@@ -821,7 +841,14 @@ class charge_distribution_surface<Lyt, false> : public Lyt
...
 
-            strg->system_energy += -(this->get_local_potential_by_index(random_element).value());
+            strg->system_energy +=
+                units::energy::electron_volt_t(units::convert<units::energy::joule, units::energy::electron_volt>(
+                    (-(*this->get_local_potential_by_index(random_element)) * units::constants::e).value()));
+
+            //            strg->system_energy =
+            //                units::energy::electron_volt_t(units::convert<units::energy::joule,
+            //                units::energy::electron_volt>(
+            //                    (total_potential * units::constants::e).value()));
```
Comment (marcelwa) [inline include/fiction/technology/charge_distribution_surface.hpp:851]: Please remove.

File: include/fiction/technology/charge_distribution_surface.hpp [hunk 16]
```diff
@@ -788,10 +808,10 @@ class charge_distribution_surface<Lyt, false> : public Lyt
                 continue;
             }
 
-            const auto dist_min =
+            const double dist_min =
```
Comment (marcelwa) [inline include/fiction/technology/charge_distribution_surface.hpp:811]: If `dist_min` represents a distance, should we not use a unit type here?

File: include/fiction/technology/physical_constants_and_ev_unit.hpp
Comment (marcelwa) [inline include/fiction/technology/physical_constants_and_ev_unit.hpp:1]: Why did you rename the file?

I see that it might have lost most of its content due to the introduction of the unit library. It doesn't really hold any physical constants anymore though. So maybe the new name isn't really fitting either. Instead, maybe you can make this into some sort of `units_utils.hpp`? Where you define all the unit literals and aliases, etc. in a single place? The `physical_constants` namespace seems not needed anymore.

File: include/fiction/technology/sidb_nm_position.hpp [hunk 17]
```diff
@@ -23,16 +23,18 @@ namespace fiction
...
+sidb_nm_position(const sidb_simulation_parameters& sp, const cell<Lyt>& c) noexcept
 {
     static_assert(has_siqad_coord_v<Lyt>, "Lyt is not based on SiQAD coordinates");
 
-    const auto x = static_cast<double>(c.x * sp.lat_a * 0.1);
-    const auto y = static_cast<double>(c.y * sp.lat_b * 0.1 + c.z * sp.lat_c * 0.1);
+    const auto x = units::length::nanometer_t(
+        units::convert<units::length::angstrom, units::length::nanometer>((c.x * sp.lat_a).value()));
+    const auto y = units::length::nanometer_t(
+        units::convert<units::length::angstrom, units::length::nanometer>((c.y * sp.lat_b + c.z * sp.lat_c).value()));
```
Comment (marcelwa) [inline include/fiction/technology/sidb_nm_position.hpp:34]: Are the `.value()` calls needed?

File: libs/CMakeLists.txt [hunk 18]
```diff
@@ -140,3 +140,7 @@ target_link_libraries(libfiction INTERFACE graph-coloring)
 set(tinyxml2_BUILD_TESTING OFF)
 add_subdirectory(tinyxml2/)
 target_link_libraries(libfiction INTERFACE tinyxml2::tinyxml2)
+
+#include units
```
Comment (marcelwa) [inline libs/CMakeLists.txt:144]:
```suggestion
# Include units
```

File: test/algorithms/simulation/sidb/minimum_energy.cpp [hunk 19]
```diff
@@ -30,11 +30,12 @@ TEMPLATE_TEST_CASE(
...
         std::vector<charge_distribution_surface<TestType>> all_lyts{};
 
-        CHECK_THAT(minimum_energy(all_lyts), Catch::Matchers::WithinAbs(std::numeric_limits<double>::max(), 0.00001));
+        CHECK_THAT(minimum_energy(all_lyts).value(),
+                   Catch::Matchers::WithinAbs(std::numeric_limits<double>::max(), 0.00001));
 
         all_lyts.push_back(charge_layout);
 
-        CHECK(std::abs(minimum_energy(all_lyts) - 0) < 0.00000001);
+        CHECK(std::abs(minimum_energy(all_lyts).value() - 0) < 0.00000001);
```
Comment (marcelwa) [inline test/algorithms/simulation/sidb/minimum_energy.cpp:38]: `units::math::abs`

File: test/algorithms/simulation/sidb/minimum_energy.cpp [hunk 20]
```diff
@@ -44,11 +45,12 @@ TEMPLATE_TEST_CASE(
...
```
Comment (marcelwa) [inline test/algorithms/simulation/sidb/minimum_energy.cpp:53]: `units::math::abs`

File: test/io/write_sqd_sim_result.cpp
Comment (marcelwa) [inline test/io/write_sqd_sim_result.cpp:1]: Due to the `using namespace fiction;` statement, the `fiction::` namespace operators can be dropped across this file.

File: test/technology/charge_distribution_surface.cpp [hunk 21]
```diff
@@ -15,41 +16,41 @@
...
+//     CHECK(has_get_charge_state_v<charge_layout>);
+//
+//     const charge_layout defect_lyt{};
+//     const charge_layout defect_lyt_from_lyt{lyt};
+//
+//     using charge_charge_layout = charge_distribution_surface<charge_layout>;
+//     CHECK(is_cell_level_layout_v<charge_charge_layout>);
+//     CHECK(has_assign_charge_state_v<charge_charge_layout>);
+//     CHECK(has_get_charge_state_v<charge_charge_layout>);
+// }
```
Comment (marcelwa) [inline test/technology/charge_distribution_surface.cpp:53]: What happened here?

File: test/technology/charge_distribution_surface.cpp [hunk 22]
```diff
@@ -70,119 +71,119 @@ TEMPLATE_TEST_CASE(
...
+    //        charge_layout.assign_charge_state({5, 6}, sidb_charge_state::NEGATIVE);
+    //
+    //        // remove previously assigned charge states
+    //        charge_layout.assign_charge_state({5, 4}, sidb_charge_state::NONE);
+    //        charge_layout.assign_charge_state({5, 5}, sidb_charge_state::NONE);
+    //        charge_layout.assign_charge_state({5, 6}, sidb_charge_state::POSITIVE);
+    //        CHECK(charge_layout.get_charge_state({5, 4}) == sidb_charge_state::NONE);
+    //        CHECK(charge_layout.get_charge_state({5, 5}) == sidb_charge_state::NONE);
+    //        CHECK(charge_layout.get_charge_state({5, 6}) == sidb_charge_state::POSITIVE);
+    //    }
```
Comment (marcelwa) [inline test/technology/charge_distribution_surface.cpp:186]: And here?

File: test/technology/charge_distribution_surface.cpp [hunk 23]
```diff
@@ -192,20 +193,20 @@ TEMPLATE_TEST_CASE(
...
+        CHECK_THAT(charge_layout.get_nm_distance_between_cells({1, 0, 0}, {1, 0, 0}).value(),
                    Catch::Matchers::WithinAbs(0.0, 0.00001));
-        CHECK_THAT(charge_layout.get_nm_distance_between_cells({0, 0, 0}, {1, 1, 1}),
-                   Catch::Matchers::WithinAbs(
-                       std::hypot(sidb_simulation_parameters{}.lat_a * 0.1,
-                                  sidb_simulation_parameters{}.lat_b * 0.1 + sidb_simulation_parameters{}.lat_c * 0.1),
-                       0.00001));
-        CHECK_THAT(charge_layout.get_nm_distance_between_cells({1, 1, 1}, {1, 1, 1}),
+        CHECK_THAT(charge_layout.get_nm_distance_between_cells({0, 0, 0}, {1, 1, 1}).value(),
+                   Catch::Matchers::WithinAbs(std::hypot(sidb_simulation_parameters{}.lat_a.value() * 0.1,
```
Comment (marcelwa) [inline test/technology/charge_distribution_surface.cpp:205]: `units::math::hypot`

File: test/technology/charge_distribution_surface.cpp [hunk 24]
```diff
@@ -2,6 +2,7 @@
 // Created by Jan Drewniok on 05.12.22.
 //
 
+#include <catch2/catch_approx.hpp>
```
Comment (marcelwa) [inline test/technology/charge_distribution_surface.cpp:5]: Sorry, I must have missed what you use this include for.

File: test/technology/charge_distribution_surface.cpp (hunk 21 above)
Comment (Drewniok) [reply test/technology/charge_distribution_surface.cpp:53]: oh, sorry! I did this to find a bug but unfortunately, I forgot to undo it. Thanks you!

File: include/fiction/algorithms/path_finding/distance.hpp [hunk 25]
```diff
@@ -71,23 +71,22 @@ template <typename Lyt, typename Dist = double>
...
     const auto pos_c1 = sidb_nm_position<Lyt>(sp, source);
     const auto pos_c2 = sidb_nm_position<Lyt>(sp, target);
 
-    const auto x = static_cast<Dist>(pos_c1.first) - static_cast<double>(pos_c2.first);
-    const auto y = static_cast<Dist>(pos_c1.second) - static_cast<double>(pos_c2.second);
+    const auto x = pos_c1.first - pos_c2.first;
+    const auto y = pos_c1.second - pos_c2.second;
 
-    return static_cast<Dist>(std::hypot(x, y));
+    return units::length::nanometer_t(std::hypot(x.value(), y.value()));
```
[the remark this answers is not in the document]
Comment (Drewniok) [reply include/fiction/algorithms/path_finding/distance.hpp:89]: That's cool! Indeed, it works!

File: include/fiction/algorithms/simulation/sidb/calculate_energy_and_state_type.hpp [hunk 26]
```diff
@@ -54,12 +54,12 @@ calculate_energy_and_state_type(const sidb_energy_distribution&
     for (const auto& [energy, occurrence] : energy_distribution)
     {
         // round the energy value to six decimal places to overcome potential rounding errors.
-        const auto energy_value = round_to_n_decimal_places(energy, 6);
+        const auto energy_value = round_to_n_decimal_places(energy.value(), 6);
```
[the remark this answers is not in the document]
Comment (Drewniok) [reply include/fiction/algorithms/simulation/sidb/calculate_energy_and_state_type.hpp:57]: yeah, good idea.

File: include/fiction/algorithms/simulation/sidb/critical_temperature.hpp [hunk 27]
```diff
@@ -123,29 +126,30 @@ struct critical_temperature_stats
     /**
      * Critical Temperature of the given layout.
      */
-    double critical_temperature{};
+    units::temperature::kelvin_t critical_temperature{};
```
[the remark this answers is not in the document]
Comment (Drewniok) [reply include/fiction/algorithms/simulation/sidb/critical_temperature.hpp:129]: sure!

File: include/fiction/algorithms/simulation/sidb/critical_temperature.hpp [hunk 28]
```diff
@@ -123,29 +126,30 @@ struct critical_temperature_stats
...
     /**
      * Number of physically valid charge configurations.
      */
     uint64_t num_valid_lyt{};
     /**
      * Energy difference between the ground state and the first (erroneous) excited state.
      */
-    double energy_between_ground_state_and_first_erroneous = std::numeric_limits<double>::infinity();
+    units::energy::electron_volt_t energy_between_ground_state_and_first_erroneous =
+        units::energy::electron_volt_t(std::numeric_limits<double>::infinity());
```
[the remark this answers is not in the document]
Comment (Drewniok) [reply include/fiction/algorithms/simulation/sidb/critical_temperature.hpp:138]: Unfortunately, this does not work. I really need to build the unit.

File: include/fiction/algorithms/simulation/sidb/is_ground_state.hpp (hunk 3 above)
Comment (Drewniok) [reply include/fiction/algorithms/simulation/sidb/is_ground_state.hpp:45]: It was an old version before. I used ``std::abs`` before we had ``round_to_n_decimal_places``.

File: include/fiction/technology/charge_distribution_surface.hpp (hunk 12 above)
Comment (Drewniok) [reply include/fiction/technology/charge_distribution_surface.hpp:596]: sorry

File: include/fiction/algorithms/simulation/sidb/occupation_probability_of_excited_states.hpp (hunk 5 above)
Comment (marcelwa) [reply include/fiction/algorithms/simulation/sidb/occupation_probability_of_excited_states.hpp:56]: You're absolutely right. My bad.

File: include/fiction/technology/charge_distribution_surface.hpp (hunk 16 above)
Comment (Drewniok) [reply include/fiction/technology/charge_distribution_surface.hpp:811]: True! Thanks!

File: include/fiction/technology/sidb_nm_position.hpp (hunk 17 above)
Comment (Drewniok) [reply include/fiction/technology/sidb_nm_position.hpp:34]: As far as I understand it, the convert function needs a double.

File: test/io/write_sqd_sim_result.cpp
Comment (Drewniok) [reply test/io/write_sqd_sim_result.cpp:1]: You are absolutely right. I think I got a warning that something was ambiguous. That was the reason, I guess, why I added ``fiction``. Anyway, it works now. Thanks!

File: docs/utils/utils.rst [hunk 29]
```diff
@@ -125,6 +124,17 @@ Array Utils
 .. doxygenfunction:: fiction::convert_array
 .. doxygenfunction:: fiction::convert_array_of_arrays
 
+Unit Utils
+-----------
+
+**Header:** ``fiction/utils/unit_utils.hpp``
+
```
Comment (Drewniok) [inline docs/utils/utils.rst:131]: @marcelwa Do you have an idea how I add the content from ``units_utils.hpp``?

Comment (marcelwa) [reply docs/utils/utils.rst:131]: For aliases, you can use `.. doxygentypedef::`. Not sure how to handle the `using ...` statements. Maybe you can find something here: https://breathe.readthedocs.io/en/latest/directives.html

Comment (marcelwa) [reply docs/utils/utils.rst:131]: Then maybe just add a bit of text here mentioning that there are these operators exposed via this file.

File: include/fiction/algorithms/path_finding/distance.hpp [hunk 30]
```diff
@@ -71,23 +71,22 @@ template <typename Lyt, typename Dist = double>
...
     const auto pos_c1 = sidb_nm_position<Lyt>(sp, source);
     const auto pos_c2 = sidb_nm_position<Lyt>(sp, target);
 
-    const auto x = static_cast<Dist>(pos_c1.first) - static_cast<double>(pos_c2.first);
-    const auto y = static_cast<Dist>(pos_c1.second) - static_cast<double>(pos_c2.second);
+    const auto x = pos_c1.first - pos_c2.first;
+    const auto y = pos_c1.second - pos_c2.second;
 
-    return static_cast<Dist>(std::hypot(x, y));
+    return {units::math::hypot(x, y)};
```
Comment (marcelwa) [inline include/fiction/algorithms/path_finding/distance.hpp:89]: Now with the usage of `units::math::hypot`, you probably don't need the braces anymore.

File: include/fiction/algorithms/simulation/sidb/critical_temperature.hpp [hunk 31]
```diff
@@ -95,9 +97,9 @@ struct critical_temperature_params
      */
     double confidence_level{0.99};
     /**
-     * Simulation stops at max_temperature (room temperature ~300 K).
+     * Simulation stops at max_temperature (~ 126 °C).
```
Comment (marcelwa) [inline include/fiction/algorithms/simulation/sidb/critical_temperature.hpp:100]:
```suggestion
     * Simulation stops at max_temperature (~ 126 °C by default).
```

File: include/fiction/algorithms/simulation/sidb/time_to_solution.hpp [hunk 32]
```diff
@@ -11,6 +11,8 @@
 #include "fiction/algorithms/simulation/sidb/quicksim.hpp"
 #include "fiction/technology/charge_distribution_surface.hpp"
 #include "fiction/traits.hpp"
+#include "fiction/utils/units_utils.hpp"
+#include "units.h"
```
Comment (marcelwa) [inline include/fiction/algorithms/simulation/sidb/time_to_solution.hpp:15]:
```suggestion
#include <units.h>
```

File: include/fiction/utils/units_utils.hpp [hunk 33]
```diff
@@ -0,0 +1,60 @@
+//
+// Created by Jan Drewniok on 24.11.22.
+//
+
+#ifndef FICTION_UNITS_UTILS_HPP
+#define FICTION_UNITS_UTILS_HPP
+
+#include "units.h"
```
Comment (marcelwa) [inline include/fiction/utils/units_utils.hpp:8]:
```suggestion
#include <units.h>
```

File: include/fiction/utils/units_utils.hpp [hunk 34]
```diff
@@ -0,0 +1,60 @@
...
+/**
+ *  Unit of the Coulomb constant (\f$ N \cdot m^{2} \cdot C^{-2} \f$).
+ */
+using coulomb_constant_unit =
+    units::unit_t<units::compound_unit<units::force::newtons, units::area::square_meter,
+                                       units::inverse<units::squared<units::charge::coulomb>>>>;
+/**
+ * The pop stability error is used for physical simulations to avoid floating-point errors.
+ */
+constexpr double POP_STABILITY_ERR = 1E-6;
```
Comment (marcelwa) [inline include/fiction/utils/units_utils.hpp:56]:
```suggestion
inline constexpr double POP_STABILITY_ERR = 1E-6;
```

File: test/algorithms/simulation/sidb/exhaustive_ground_state_simulation.cpp [hunk 35]
```diff
@@ -5,12 +5,13 @@
 #include <catch2/catch_template_test_macros.hpp>
 #include <catch2/matchers/catch_matchers_floating_point.hpp>
 
+#include "fiction/utils/units_utils.hpp"
```
Comment (marcelwa) [inline test/algorithms/simulation/sidb/exhaustive_ground_state_simulation.cpp:8]:
```suggestion
#include <fiction/utils/units_utils.hpp>
```

File: test/algorithms/simulation/sidb/quicksim.cpp [hunk 36]
```diff
@@ -5,12 +5,13 @@
 #include <catch2/catch_template_test_macros.hpp>
 #include <catch2/matchers/catch_matchers_floating_point.hpp>
 
+#include "fiction/utils/units_utils.hpp"
```
Comment (marcelwa) [inline test/algorithms/simulation/sidb/quicksim.cpp:8]:
```suggestion
#include <fiction/utils/units_utils.hpp>
```

**Review by marcelwa — CHANGES_REQUESTED** [review]
Looks really good. I spotted an inconsistency here and there. Besides that, and the fact that Windows CI is not passing for some reason, this seems done.

File: docs/utils/utils.rst [hunk 37]
```diff
@@ -125,6 +124,15 @@ Array Utils
...
+.. doxygentypedef:: coulomb_constant_unit
+
+There are several operators exposed in this file for convenient use of units.
+The following notation is possible: value_angstrom, value_eV, value_K, value_nm, value_s, value_V (e.g. 40_K, 40.1_nm).
```
Comment (marcelwa) [inline docs/utils/utils.rst:135]:
```suggestion
There are several suffix operators exposed in this file for convenient use of units, namely

- ``_angstrom``
- ``_eV``
- ``_K``
- ``_nm``
- ``_s``
 - ``_V``
 
This enables the use of units as follows: e.g., ``40_K``, ``40.1_nm``.
```

File: docs/utils/utils.rst [hunk 38]
```diff
@@ -149,11 +157,11 @@ One can use the following macros to specify the execution policy for parallel ST
 platform-independent way::
 
     std::for_each(FICTION_EXECUTION_POLICY_PAR v.begin(), v.end(), lambda);
-    //                                        ^ note the missing comma
+//                                        ^ note the missing comma
 
 If parallelism or execution policies are not available, this will expand to::
 
-    std::for_each(v.begin(), v.end(), lambda);
+                                                            std::for_each(v.begin(), v.end(), lambda);
```
Comment (marcelwa) [inline docs/utils/utils.rst:164]: Please revert these changes.

File: include/fiction/algorithms/simulation/sidb/time_to_solution.hpp [hunk 39]
```diff
@@ -126,9 +130,9 @@ void sim_acc_tts(const Lyt& lyt, const quicksim_params& quicksim_params, time_to
         tts = (single_runtime * std::log(1.0 - confidence_level) / std::log(1.0 - acc));
     }
 
-    st.time_to_solution    = tts;
+    st.time_to_solution    = units::time::second_t(tts);
     st.acc                 = acc * 100;
-    st.mean_single_runtime = single_runtime;
+    st.mean_single_runtime = units::time::second_t(single_runtime);
```
Comment (marcelwa) [inline include/fiction/algorithms/simulation/sidb/time_to_solution.hpp:135]: Same here.

File: include/fiction/algorithms/simulation/sidb/time_to_solution.hpp [hunk 40]
```diff
@@ -82,7 +85,8 @@ void sim_acc_tts(const Lyt& lyt, const quicksim_params& quicksim_params, time_to
     const auto simulation_results_exgs = exhaustive_ground_state_simulation(lyt, quicksim_params.phys_params);
 
     time_to_solution_stats st{};
-    st.single_runtime_exhaustive = mockturtle::to_seconds(simulation_results_exgs.simulation_runtime);
+    st.single_runtime_exhaustive =
+        units::time::second_t(mockturtle::to_seconds(simulation_results_exgs.simulation_runtime));
```
Comment (marcelwa) [inline include/fiction/algorithms/simulation/sidb/time_to_solution.hpp:89]: Can you use braced initialization here, please? It is generally preferrable to avoid accidentally declaring a function.

File: include/fiction/algorithms/simulation/sidb/time_to_solution.hpp (hunk 39 above)
Comment (marcelwa) [inline include/fiction/algorithms/simulation/sidb/time_to_solution.hpp:135]: Same here.

File: include/fiction/utils/units_utils.hpp [hunk 41]
```diff
@@ -0,0 +1,60 @@
...
+/**
+ *  Unit of the Coulomb constant (\f$ N \cdot m^{2} \cdot C^{-2} \f$).
+ */
+using coulomb_constant_unit =
+    units::unit_t<units::compound_unit<units::force::newtons, units::area::square_meter,
+                                       units::inverse<units::squared<units::charge::coulomb>>>>;
+/**
+ * The pop stability error is used for physical simulations to avoid floating-point errors.
+ */
+inline constexpr double POP_STABILITY_ERR = 1E-6;
```
Comment (marcelwa) [inline include/fiction/utils/units_utils.hpp:56]: Does this constant have RST documentation already?

**Review by marcelwa — CHANGES_REQUESTED** [review]
Looks great! Except that the Windows CI is still not compiling. That needs to be addressed before this PR can be merged.

File: include/fiction/utils/math_utils.hpp [hunk 42]
```diff
@@ -14,19 +17,33 @@ namespace fiction
...
-    return std::round(number * factor) / factor;
+    static_assert(std::is_arithmetic_v<T> || units::traits::is_temperature_unit<T>::value ||
+                      units::traits::is_length_unit<T>::value || units::traits::is_voltage_unit<T>::value ||
+                      units::traits::is_energy_unit<T>::value,
+                  "T is neither a number type nor a unit type");
+
+    if constexpr (std::is_arithmetic_v<T>)
+    {
+        const T factor = std::pow(10, n);
+        return std::round(number * factor) / factor;
```
[the remark this answers is not in the document]
Comment (marcelwa) [reply include/fiction/utils/math_utils.hpp:39]: @Drewniok when this remaining comment is addressed, this PR can be merged :)

File: include/fiction/algorithms/simulation/sidb/is_ground_state.hpp [hunk 43]
```diff
@@ -42,8 +44,8 @@ template <typename Lyt>
     const auto min_energy_exact  = minimum_energy(exhaustive_results.charge_distributions);
     const auto min_energy_new_ap = minimum_energy(quicksim_results.charge_distributions);
 
-    return round_to_n_decimal_places(std::abs(min_energy_exact - min_energy_new_ap),
-                                     physical_constants::POP_STABILITY_ERR) == 0;
+    return round_to_n_decimal_places(units::math::abs(min_energy_exact - min_energy_new_ap), POP_STABILITY_ERR)
```
[the remark this answers is not in the document]
Comment (Drewniok) [reply include/fiction/algorithms/simulation/sidb/is_ground_state.hpp:47]: I see. I did a mistake here. I will fix it.

File: include/fiction/utils/math_utils.hpp [hunk 44]
```diff
@@ -14,19 +17,33 @@ namespace fiction
...
-    const T factor = std::pow(10, n);
-    return std::round(number * factor) / factor;
+    static_assert(std::is_arithmetic_v<T> || units::traits::is_temperature_unit<T>::value ||
+                      units::traits::is_length_unit<T>::value || units::traits::is_voltage_unit<T>::value ||
+                      units::traits::is_energy_unit<T>::value,
+                  "T is neither a number type nor a unit type");
+
+    if constexpr (std::is_arithmetic_v<T>)
+    {
+        const T factor = std::pow(10, n);
```
Comment (marcelwa) [inline include/fiction/utils/math_utils.hpp:38]:
```suggestion
        const T factor = std::pow(10.0, n);
```

File: include/fiction/utils/math_utils.hpp [hunk 45]
```diff
@@ -14,19 +17,33 @@ namespace fiction
...
+        return std::round(number * factor) / static_cast<double>(factor);
+    }
+    else if constexpr (units::traits::is_temperature_unit<T>::value || units::traits::is_length_unit<T>::value ||
+                       units::traits::is_voltage_unit<T>::value || units::traits::is_energy_unit<T>::value)
+    {
+        const double factor = std::pow(10, n);
```
Comment (marcelwa) [inline include/fiction/utils/math_utils.hpp:44]:
```suggestion
        const double factor = std::pow(10.0, n);
```

**Review by marcelwa — CHANGES_REQUESTED** [review]

**Review by marcelwa — APPROVED** [review]
